### DISCUSSION TITLE
Missing viewState from a FrontStage

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#3593] [Q&A] [answered]
What should be done when an imodel doesn't contain a schedule and a viewState is not found as a result ?

A case resulting from this [sandbox example](https://www.itwinjs.org/sandboxes/NancyMcCall/Default%20View%20Frontstage).

```typescript
let viewState;
const defaultViewId = await iModel?.views?.queryDefaultViewId();
if (defaultViewId && Id64.isValidId64(defaultViewId)) {
    viewState = await iModel?.views.load(defaultViewId);
    viewState && UiFramework.setDefaultViewState(viewState.clone());
}
```

If `defaultViewId` is 0 or undefined then the condition will not be true and the `viewState` will not be set, resulting in this:

![image](https://user-images.githubusercontent.com/17766221/167258351-c8e54625-99f2-4a23-b6f5-773d8df3e4fb.png)

I think there might be some missing steps in the sandbox code to implement a proper frontstage ? This is backed by the existence of an extra measuring tool on top of the selection tool (which should indicate a problem):

![image](https://user-images.githubusercontent.com/17766221/167258445-b36d59ac-daaf-4375-886a-cb6373e813ee.png)

I don't want to create a custom viewstate each time one is not defined, like [this](https://github.com/iTwin/itwinjs-core/blob/c2700b08588f8117d3c1c860817c23c3a11f35b7/test-apps/ui-test-app/src/frontend/appui/frontstages/FrontstageUi2.tsx#L59) 

I also noticed that it's possible to fetch every available view and and get a defaultViewId from the first one for example:

```typescript
await iModel?.views?.getViewList({}).then((result)=> defaultViewId = result[0].id);
```

But this ends up resulting in undesired views saved from the BIM Software (Synchro/Revit..) and is not the same as a default viewState created automatically.

I just want to render a timeline if there's any, with a default viewState. I feel I'm going in circles trying to do something very simple.

**Comment by pmconne** [comment, marked as answer]
Hi @oussamabonnor1,

I'm a little confused as to what you are actually asking. Tell me if I get it right: you want to open an iModel, find a schedule script in it, and if found apply that script to your viewport.

**Reply by oussamabonnor1** [reply]
Yes, at Evercam we deal with many iModels, some of them have schedules and others dont. I want to render a timeline for those who have a schedule and for those who dont, just render the simple model.

Nothing fancy actually.

The problem is that it was suggested to me to use a custom frontstage to use the schedule but now the non-schedule having iModels are not being displayed.

**Reply by pmconne** [reply]
That sandbox was provided to illustrate how to solve a specific problem. It's not intended to be a full-scale app - it's assuming for simplicity that a default view exists.
Which part(s) do you need assistance with?
- Locating a timeline in the iModel
- Creating a view if no default view exists
- Associating a timeline with a view that does not have a timeline
- ...?

If at all possible please show your code.

**Reply by oussamabonnor1** [reply]
Creating a view if no default view exists. That one, thanks.
My code is the same as the sandbox, I didn't change anything yet.

Could a more detailed approach to implemented a frontstage be added later ? That'd be very helpful!

**Reply by pmconne** [reply]
Have you tried [ViewCreator3d](https://www.itwinjs.org/reference/core-frontend/views/viewcreator3d/)?
I can't help with frontstages - I'm not a UI guy.

**Reply by oussamabonnor1** [reply]
@pmconne lunch's on me if we ever meet my friend!
